tests/drivers/clock_control/clock_control_api/src/test_clock_control.c

```diff
@@ -0,0 +1,318 @@
+	zassert_true(status != CLOCK_CONTROL_STATUS_OFF,
+			"%s: Unexpected status (%d)", dev_name, status);
+
+	for (int i = 0; i < users; i++) {
+		err = clock_control_off(dev, 0);
+		zassert_equal(0, err, "%s: Unexpected err (%d)", dev_name, err);
+	}
+
+	status = clock_control_get_status(dev, NULL);
+	zassert_true(status == CLOCK_CONTROL_STATUS_OFF,
+			"%s: Unexpected status (%d)", dev_name, status);
+}
+
+static void test_multiple_users(void)
+{
+	test_all_instances(test_multiple_users_instance, NULL);
+}
+
+static bool async_capable(const char *dev_name)
+{
+	struct device *dev = device_get_binding(dev_name);
+
+	if (clock_control_async_on(dev, 0, NULL) != 0) {
+		return false;
+	}
+
+	clock_control_off(dev, 0);
+
+	return true;
+}
+
+/*
+ * Test checks that callbacks are called after clock is started.
+ */
+static void clock_on_callback1(struct device *dev, void *user_data)
+{
+	bool *executed = (bool *)user_data;
+
+	*executed = true;
+}
+
+static void clock_on_callback2(struct device *dev, void *user_data)
+{
+	bool *executed = (bool *)user_data;
+
+	*executed = true;
+}
+
+static void test_async_on_instance(const char *dev_name, u32_t startup_us)
+{
+	struct device *dev = device_get_binding(dev_name);
+	enum clock_control_status status;
+	int err;
+	bool executed1 = false;
+	bool executed2 = false;
+	struct clock_control_async_data data1 = {
+		.cb = clock_on_callback1,
+		.user_data = &executed1
+	};
+	struct clock_control_async_data data2 = {
+		.cb = clock_on_callback2,
+		.user_data = &executed2
+	};
+
+	status = clock_control_get_status(dev, NULL);
+	zassert_equal(CLOCK_CONTROL_STATUS_OFF, status,
+			"%s: Unexpected status (%d)", dev_name, status);
+
+	err = clock_control_async_on(dev, 0, &data1);
+	zassert_equal(0, err, "%s: Unexpected err (%d)", dev_name, err);
+
+	err = clock_control_async_on(dev, 0, &data2);
+	zassert_equal(0, err, "%s: Unexpected err (%d)", dev_name, err);
+
+	/* wait for clock started. */
+	k_busy_wait(startup_us*1.5);
+
+	zassert_true(executed1, "%s: Expected flag to be true", dev_name);
+	zassert_true(executed2, "%s: Expected flag to be true", dev_name);
+}
+
+static void test_async_on(void)
+{
+	test_all_instances(test_async_on_instance, async_capable);
+}
+
+/*
+ * Test checks that when asynchronous clock enabling is scheduled but clock
+ * is disabled before being started then callback is never called.
+ */
+static void test_async_on_stopped_on_instance(const char *dev_name,
+						u32_t startup_us)
+{
+	struct device *dev = device_get_binding(dev_name);
+	enum clock_control_status status;
+	int err;
+	int key;
+	bool executed1 = false;
+	struct clock_control_async_data data1 = {
+		.cb = clock_on_callback1,
+		.user_data = &executed1
+	};
+
+	status = clock_control_get_status(dev, NULL);
+	zassert_equal(CLOCK_CONTROL_STATUS_OFF, status,
+			"%s: Unexpected status (%d)", dev_name, status);
+
+	/* lock to prevent clock interrupt for fast starting clocks.*/
+	key = irq_lock();
```

Comment:
This will only prevent handling of the interrupt. If a clock starts fast enough, it can still generate the interrupt before `clock_control_off` is called. In such case the interrupt will be handled (and the callback executed) right after `irq_unlock`.

Reply:
that's why clearing started event after stopping the clock. Not sure how it will be done in other platforms but driver should ensure that interrupt will not be generated after stopping the clock before it got started.